README.md
# gifsockets-middleware [![Build status](https://travis-ci.org/twolfson/gifsockets-middleware.png?branch=master)](https://travis-ci.org/twolfson/gifsockets-middleware)

Set of HTTP middlewares for [gifsockets][]

This is part of the [gifsockets project][]; it acts as a plug and play middleware that is used in the [demo][gifsockets project] but can be re-used anywhere.

[gifsockets]: https://github.com/twolfson/gifsockets
[gifsockets project]: https://github.com/twolfson/gifsockets-server

## Getting Started
Install the module with: `npm install gifsockets-middleware`

```javascript
var GifsocketMiddleware = require('gifsockets-middleware');
var middlewares = GifsocketMiddleware({width: 200, height: 200});
var express = require('express');
var app = express();

// middlewares returns an object containing 4 middlewares
// `openImage` writes the beginning of a .gif and leaves `res` open
app.get('/image.gif', middlewares.openImage);

// `writePixelsToImages` writes a new frame to all open `res` from openImage
var bodyParser = express.bodyParser();
app.post('/image/pixels', bodyParser, middlewares.writePixelsToImages);

// `writeTextToImages` accepts a string of text and writes a new frame
// This requires running `phantomjs-pixel-server`
app.post('/image/text', bodyParser, middlewares.writeTextToImages);

// `closeOpenImages` closes all active images opened by `openImage`
app.post('/image/close', middlewares.closeOpenImages);

// If you want to load a specific middleware, you can do so
var openImageMiddleware = require('gifsockets-middleware/lib/middlewares/open-image');
var openImage = openImageMiddleware(gifsocket);
// `openImage` has the same behavior as that returned from `GifsocketMiddleware`
```

## Documentation
`gifsockets-middleware` returns `GifsocketMiddleware` as its `module.exports`

### `GifsocketMiddleware(options)`
Function that generates an object of middlewares for `gifsockets`

- options `Object`
    - width `Number` Width of the output GIF/`gifsocket`
    - height `Number` Height of the output GIF/`gifsocket`
- Returns an object containing
    - `openImage` middleware
    - `writePixelsToImages` middleware
    - `writeTextToImages` middleware
    - `closeOpenImages` middleware

### `openImage` middleware
Middleware that will maintain an open connection such that it can write additional GIF frames.

Function signature is `function (req, res, next) {}`

This does not expect any information on `req`/`res` and will not callback to `next`.

### `writePixelsToImages` middleware
Middleware that will write a new GIF frame with the provided pixels.

Function signature is `function (req, res, next) {}`

If `req.rgbPixels` exists, we will draw a GIF frame with the pixel values.

> `req.rgbPixels` is expected to be an stringified array of rgb pixels; `[0, 1, 2, 3, 4, 5]` is 2 pixels with `r: 0, g: 1, b: 2` and `r: 3, g: 4, b: 5`

If `req.rgbPixels` is not found, we look for `req.rgbaPixels` or `req.body`. If either of these is found, we will draw a GIF frame with the pixel values.

> `req.rgbaPixels`/`req.body` is expected to be an stringified array of rgba pixels; `[0, 1, 2, 3, 4, 5, 6, 7]` is 2 pixels with `r: 0, g: 1, b: 2, a: 3` and `r: 4, g: 5, b: 6, a: 7`

This will reply with a `204` when it is complete.

### `writeTextToImages` middleware
**This depends on [phantomjs-pixel-server][] running as another process.**

**Please read the documentation: https://github.com/twolfson/phantomjs-pixel-server**

[phantomjs-pixel-server]: https://github.com/twolfson/phantomjs-pixel-server

Middleware that will create and write a new GIF frame with the provided text parameters.

Function signature is `function (req, res, next) {}`

If `req.body` is provided, we will parse it via [`querystring.parse`][] and use the following parameters:

- text `String` - Text to write to the canvas
- background `String` - [CSS color][] background for the canvas (default "#000000")
  - Any valid [CSS color][] is accepted, under the hood we are using [`fillStyle`][]
- foreground `String` - [CSS color][] for text on the canvas (default "#BADA55")
  - Any valid [CSS color][] is accepted, under the hood we are using [`fillStyle`][]
- font-size `Number` - Size of the font to draw (default: 30)
- font-family `String` - Font family to use on the canvas (default: "Impact")

[`querystring.parse`]: http://nodejs.org/api/querystring.html
[CSS color]: https://developer.mozilla.org/en-US/docs/Web/CSS/color
[`fillStyle`]: http://www.w3.org/TR/2dcontext/#fill-and-stroke-styles

This will reply with a `204` when it is complete.

### `closeOpenImages` middleware
Middleware that iterates over open connections, writes GIF footer, and closes connection.

Function signature is `function (req, res, next) {}`

This will reply with a `204` when it is complete.

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint via [grunt](https://github.com/gruntjs/grunt) and test via `npm test`.

## Donating
Support this project and [others by twolfson][gittip] via [gittip][].

[![Support via Gittip][gittip-badge]][gittip]

[gittip-badge]: https://rawgithub.com/twolfson/gittip-badge/master/dist/gittip.png
[gittip]: https://www.gittip.com/twolfson/

## Unlicense
As of Nov 20 2013, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE


CHANGELOG.md
# gifsockets-middleware changelog
0.1.0 - Initial release


test/utils/image.js
var fs = require('fs');
var getPixels = require('get-pixels');

exports.loadRgbaImage = function (filename) {
  before(function loadImage (done) {
    var that = this;
    getPixels(__dirname + '/../test-files/' + filename, function (err, pixels) {
      if (err) {
        return done(err);
      }
      that.rgbaPixels = pixels.data;
      done();
    });
  });
};

exports.debug = function (filename) {
  if (process.env.DEBUG_TEST) {
    before(function saveDebugImage () {
      try { fs.mkdirSync(__dirname + '/../actual-files/'); } catch (e) {}
      fs.writeFileSync(__dirname + '/../actual-files/' + filename, this.gifData, 'binary');
    });
  }

  if (false && process.env.TRAVIS) {
    before(function outputDebugImage () {
      console.log(encodeURIComponent(this.gifData));
      // Counter to it:
      // var fs = require('fs');
      // var data = "<%= data %>";
      // fs.writeFileSync('tmp.gif', decodeURIComponent(data), 'binary');
    });
  }
};

test/utils/server.js
var spawn = require('child_process').spawn;
var pixelServerPath = require.resolve('phantomjs-pixel-server');

exports.runPixelServer = function () {
  before(function startPhantomPixelServer (done) {
    this._phantomServer = spawn('phantomjs', [pixelServerPath], {stdio: [0, 1, 2]});
    setTimeout(done, 1000);
  });
  after(function (done) {
    this._phantomServer.kill();
    this._phantomServer.on('exit', function (code, signal) {
      done();
    });
  });
};


test/gifsockets-middleware_test.js
var assert = require('assert');
var fs = require('fs');
var http = require('http');
var rawBody = require('raw-body');
var request = require('request');
var imageUtils = require('./utils/image');
var serverUtils = require('./utils/server');
var GifsocketsMiddleware = require('../');

function openImage() {
  before(function (done) {
    var that = this;
    this.gifData = '';
    var req = request({
      url: 'http://localhost:8050/',
      encoding: 'binary'
    });
    req.on('error', function (err) {
      done(err);
    });
    req.on('response', function (res) {
      res.on('data', function (buff) {
        that.gifData += buff;
      });
      that.gifRes = res;
      done();
    });
  });
}

function closeImage() {
  before(function (done) {
    this.gifRes.on('end', done);
    request({
      url: 'http://localhost:8050/',
      method: 'DELETE'
    });
  });
}

describe('A server using gifsockets-middleware', function () {
  serverUtils.runPixelServer();
  before(function startGifsocketsMiddleware (done) {
    // Create a set of middlewares and server
    var middlewares = GifsocketsMiddleware({width: 10, height: 10});
    var app = http.createServer(function (req, res) {
      // DEV: We are intentionally not using express to very it works at any level
      if (req.method === 'GET') {
        middlewares.openImage(req, res);
      } else if (req.method === 'DELETE') {
        middlewares.closeOpenImages(req, res);
      } else {
        // Parse the body
        rawBody(req, function (err, buff) {
          if (err) {
            throw err;
          }

          // Save the buffer to the request
          req.body = buff;

          // Forward the request appropriately
          if (req.method === 'POST') {
            middlewares.writeJsonToImages(req, res);
          // DEV: This is a catch-all case but iddealy PUT
          } else {
            middlewares.writeTextToImages(req, res);
          }
        });
      }
    });
    app.listen(8050);

    // Save the app for later
    this.app = app;
    setTimeout(done, 100);
  });
  after(function stopGifsocketsMiddelware (done) {
    this.app.close(done);
  });

  describe.only('receiving a request', function () {
    openImage();
    before(function saveGifData () {
      this._beforeFrameData = this.gifData;
    });
    before(function writeNewFrame (done) {
      this.timeout(5000);
      request({
        url: 'http://localhost:8050/',
        method: 'PUT',
        form: {
          text: 'Hello',
          'font-family': 'Arial',
          'font-size': '2'
        }
      }, done);
    });

    it('receives a new frame', function () {
      assert.notEqual(this._beforeFrameData, this.gifData);
    });

    describe('and closing the image', function () {
      closeImage();
      imageUtils.debug('text.gif');

      it('creates a GIF image', function () {
        // TODO: We might be able to use image magick for a fuzzy match
        var expectedImages = ['text.gif', 'text2.gif'];
        var matchedAnImage = false;
        var i = expectedImages.length;
        while (i--) {
          var expectedImg = fs.readFileSync(__dirname + '/expected-files/' + expectedImages[i], 'binary');
          if (expectedImg === this.gifData) {
            matchedAnImage = true;
            break;
          }
        }
        assert(matchedAnImage);
      });
    });
  });

  describe('writing a JSON pixel frame', function () {
    openImage();
    imageUtils.loadRgbaImage('checkerboard.png');
    before(function writeJsonFrame (done) {
      request({
        url: 'http://localhost:8050/',
        method: 'POST',
        body: JSON.stringify([].slice.call(this.rgbaPixels))
      }, done);
    });
    closeImage();
    imageUtils.debug('json.gif');

    it('creates the image as a GIF', function () {
      var expectedImg = fs.readFileSync(__dirname + '/expected-files/json.gif', 'binary');
      assert.strictEqual(this.gifData, expectedImg);
    });
  });
});


lib/middlewares/write-json-to-images.js
var TextCanvas = require('../../lib/text-canvas');

module.exports = function writeJsonToImagesMiddleware (gifsocket) {
  return function writeJsonToImages (req, res, next) {
    // By default, try to find rgbPixels
    var pixelJson = req.rgbPixels;
    var method = 'writeRgbFrame';

    // Fallback to rgbaPixels
    if (!pixelJson) {
      pixelJson = req.rgbaPixels || req.body;
      method = req.drawMethod || 'writeRgbaFrame';
    }

    // If there were no pixelJson, callback
    if (!pixelJson) {
      return next();
    }

    var pixels;
    try {
      pixels = JSON.parse(pixelJson);
    } catch (e) {
      return next(new Error('JSON could not be parsed'));
    }

    // Draw the pixels using the implied method
    gifsocket[method](pixels, function wroteJsonFrame () {
      // Send a no content response
      res.writeHead(204);
      res.end();
    });
  };
};

lib/middlewares/write-text-to-images.js
var qs = require('querystring');
var TextCanvas = require('../text-canvas');

module.exports = function writTextToImagesMiddleware (gifsocket) {
  return function writTextToImages (req, res, next) {
    // Get our content to encode
    var body = req.body;

    // If there is no body, callback
    if (!body) {
      return next();
    }

    // Generate a new GIF to encode
    var query = qs.parse(body.toString());
    var textCanvas = new TextCanvas(gifsocket);
    textCanvas.getTextFrameData(query, function receiveTextFrameData (err, rawData) {
      if (err) {
        return next(err);
      }

      var rgbPixels = textCanvas.decodeStringImage(rawData);
      gifsocket.writeRgbFrame(rgbPixels, function wroteTextFrame () {
        // Send a no content response
        res.writeHead(204);
        res.end();
      });
    });
  };
};

lib/text-canvas/canvas-to-pixels/phantomjs-pixel-server.js
var assert = require('assert');
var request = require('request');
function noop() {}

module.exports = function getImageInfo (params, cb) {
  assert(params.js, 'JS not provided');
  assert(params.width, 'Width not provided');
  assert(params.height, 'Height not provided');

  // TODO: Update params.js to function-to-string

  // Stringify our argument for phantomjs
  var arg = JSON.stringify(params),
      encodedArg = encodeURIComponent(arg);

  request({
    url: 'http://localhost:9090/',
    method: 'POST',
    headers: {
      // DEV: PhantomJS looks for Proper-Case headers, request is lower-case =(
      'Content-Length': encodedArg.length
    },
    body: encodedArg
  }, function handlePhantomResponse(err, res, body) {
    // If there was an error, callback with it
    if (err) {
      return cb(err);
    }

    // If there was a bad status code, callback with a good message
    if (res.statusCode < 200 || res.statusCode >= 300) {
      return cb(new Error('PhantomJS server responded with "' + res.statusCode + '" status code. ' + body));
    }

    // Otherwise, callback with the body
    cb(null, body);
  });
};


lib/text-canvas/index.js
var assert = require('assert');
var functionToString = require('function-to-string');

var imageInfo = require('./canvas-to-pixels/phantomjs-pixel-server');
function noop() {}

function drawImage(canvas, cb) {
  var context = canvas.getContext('2d');
  var fontSize = parseInt("FONT_SIZE", 0);
  var lineHeight = Math.ceil(1.61803398875 * fontSize);
  context.font = fontSize + 'px ' + "FONT_FAMILY";
  context.fillStyle = "BACKGROUND";
  context.fillRect(0, 0, WIDTH, HEIGHT);
  context.fillStyle = "FOREGROUND";
  var text = "TEXT";
  var lines = text.split('\n');
  lines.forEach(function (line, i) {
    context.fillText(line, lineHeight / 2, (i * lineHeight) + fontSize + (lineHeight / 2));
  });
  cb(null);
}
var drawImageInfo = functionToString(drawImage);

function TextCanvas(options) {
  // Assert we have a height and width
  var width = options.width;
  var height = options.height;
  assert(width, 'Width was not provided for TextCanvas');
  assert(height, 'Height was not provided for TextCanvas');

  // Save options
  this.width = width;
  this.height = height;
}
TextCanvas.prototype = {
  decodeStringImage: function (data) {
    // Adjusted version of removeAlphaChannel
    var w = this.width;
    var h = this.height;
    var pixels = new Uint8Array(w * h * 3);

    var count = 0;

    for (var i = 0; i < h; i++) {
      for (var j = 0; j < w; j++) {
        var b = (i * w * 4) + j * 4;
        pixels[count++] = data.charCodeAt(b);
        pixels[count++] = data.charCodeAt(b+1);
        pixels[count++] = data.charCodeAt(b+2);
      }
    }

    return pixels;
  },
  getTextFrameData: function (params, cb) {
    cb = cb || noop;
    var fnBody = drawImageInfo.body
                   // Ain't no injection like a JavaScript eval injection
                   .replace('"FONT_SIZE"', JSON.stringify(params['font-size'] || '30'))
                   .replace('"FONT_FAMILY"', JSON.stringify(params['font-family'] || 'Impact'))
                   .replace('"TEXT"', JSON.stringify(params.text))
                   .replace('"BACKGROUND"', JSON.stringify(params.background || '#000000'))
                   .replace('"FOREGROUND"', JSON.stringify(params.foreground || '#BADA55'))
                   .replace('HEIGHT', this.height)
                   .replace('WIDTH', this.width);

    // TODO: Image info will operate on a switch at server start

    imageInfo({
      width: this.width,
      height: this.height,
      js: {
        params: drawImageInfo.params,
        body: fnBody
      },
      responseType: 'string'
    }, function parseFrameInfo (err, unparsedData) {
      if (err) {
        return cb(err);
      }

      cb(null, unparsedData);
    });
  }
};

module.exports = TextCanvas;